woocommerce/woocommerce
# Address possible metadata duplication in 8.1 (backport)

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




This PR applies a subset of the changes from #40129, which -among other things-, prevent duplication of metadata when using the `save_meta_data()` order method, which is common with 3rd party code (including popular plugins such as WC Subscriptions).




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure both HPOS and sync are enabled in WooCommerce > Settings > Advanced > Features.
1. Run the following PHP snippet via WP-CLI (for example):
   ```php
   $order = wc_create_order();
   $order->add_meta_data( 'this_gets_duplicated', 'for reals' );
   $order->save_meta_data();
   
   echo $order->get_id();
   ```
   Take note of the ID that gets printed to the screen.
1. Make sure that metadata `this_gets_duplicated` appears only once in the output of the following WP-CLI commands:
   1. `wp db query "SELECT * FROM $(wp db prefix)wc_orders_meta WHERE order_id = ORDER_ID"`
   1. `wp post meta list ORDER_ID`
   
   Replace `ORDER_ID` with the ID from step 2.

In 8.1 (without this fix), you'll see that steps 3 shows you the metadata duplication.




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
